tensorhive/core/services/ProtectionService.py

```diff
@@ -174,8 +185,20 @@ def do_run(self):
 
             # 4. Execute handler's behaviour on unauthorized ttys
             if len(unauthorized_sessions) > 0:
-                for handler in self.violation_handlers:
-                    handler.trigger_action(node_connection, unauthorized_sessions)
+                self.violation_handlers[0].trigger_action(node_connection, unauthorized_sessions)
+
+            # FIXME Temporary solution, refactor
+            # Only reservation owner can use GPU
+            for intruder in unpriviliged_gpu_process_owners:
+                violation_data = {
+                    'INTRUDER_USERNAME': intruder,
+                    'RESERVATION_OWNER_USERNAME': username,
+                    'RESERVATION_OWNER_EMAIL': user.email,
+                    'RESERVATION_END': reservation.ends_at,
+                    'UUID': uuid,
+                    'HOSTNAME': hostname
+                }
+                self.violation_handlers[1].trigger_action(violation_data)
```

Comment:
If i change notify_on_pty in config.ini, there is no violation_handlers[1].

BTW shouldn't the trigger_action method have the same interface for all handlers? If the parameters differ, maybe we could use a map of params?

Reply:
I was thinking exactly the same (about common interface), but it turned out that my initial implementation was just bad (it requires ssh `connections` argument which is useless for mailer - short-sighted vision).

And because I didn't want to screw up current code.... that's the workaround which I'm feeling kinda ashamed for.

I'll try to make it workin'. :cactus:

Reply:
It took me a while, but yeah... there it is (refactor was necessary indeed, see changes in `MessageSendingBehaviour`):
![image](https://user-images.githubusercontent.com/12485656/51948336-64350980-2428-11e9-8f37-f53b1e12348b.png)